Rewrite handler method argument and return type resolving

xhaggi opened the pull request:
As discussed in #127, this introduces some new Views and rewrites `HtmxResponse` for use as a method argument in controllers. This way the code is more aligned with Spring MVC. I also did some refactorings and cleanups in this PR.

The following Views are supported:

**HtmxView**
This View combines a list of `ModelAndView` and is a replacement for views from `HtmxResponse`.

**HtmxRedirectView**
This View is a subclass of `RedirectView` and therefore supports all features such as Flash attributes, exposing model attributes, context relative url etc. that `RedirectView` supports. It is a replacement for `HtmxResponse#redirect`.

**HtmxLocationRedirectView**
Same as `HtmxRedirectView` but for `HX-Location`.

**HtmxRefreshView**
This is a replacement for `HtmxResponse#refresh`.

checketts commented on htmx-spring-boot/src/main/java/io/github/wimdeblauwe/htmx/spring/boot/mvc/HtmxResponseHandlerMethodArgumentResolver.java:
```diff
@@ -1,26 +1,38 @@
+
+        if (parameter.getParameterType() == HtmxResponse.class) {
+            var htmxResponse = new HtmxResponse();
+            request.setAttribute(RequestContextUtils.HTMX_RESPONSE_CONTEXT_ATTRIBUTE, htmxResponse);
```
I lie this approach better than using the `Model`. Thanks for making that change.

checketts commented on htmx-spring-boot/src/main/java/io/github/wimdeblauwe/htmx/spring/boot/mvc/HtmxViewMethodReturnValueHandler.java:
```diff
@@ -0,0 +1,89 @@
+                                  ModelAndViewContainer mavContainer,
+                                  NativeWebRequest webRequest) throws Exception {
+
+        if (returnValue instanceof HtmxView htmxView) {
```
This `instanceof` assignment requires Java 14, right? I like it!

xhaggi replied:
It's called "Pattern Matching for instanceof" and available since Java 16.

https://openjdk.org/jeps/394

wimdeblauwe commented on htmx-spring-boot/src/main/java/io/github/wimdeblauwe/htmx/spring/boot/mvc/HtmxLocationRedirectView.java:
```diff
@@ -0,0 +1,101 @@
+ *
+ * @see <a href="https://htmx.org/headers/hx-location/">HX-Location Response Header</a>
+ */
+public class HtmxLocationRedirectView extends RedirectView {
```
I am a bit confused about this class. Why do we have all the fields to set htmx response headers in here? Shouldn't users use the `HtmxResponse` as a method parameter now?

xhaggi replied:
Let me explain this in a little more detail. I have decided to introduce derivatives of `RedirectView` to set the HX-Redirect and HX-Location headers, as htmx does not process a response body when these headers are encountered (see [source code](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L4634-L4663)). In these cases, htmx behaves like the browser and only evaluates these two headers. And as we also want to use all the functions of a `RedirectView` for an htmx redirect in Spring MVC, this simply made sense IMO. The situation is similar with `HX-Refresh`, where no response body is processed by htmx either. But in this case htmx only expects a Boolean value, so this view does not need to implement anything else.

Let me show you a real world example: A controller mapping that validates supplied form data, forward to the form on binding errors and redirect to the user list user on success. If you don't want to do anything other than this flow, you never need to inject `HtmxResponse` as a method argument. In my example, however, I want to trigger a client-side event on success.

```java
@HxRequest
@PostMapping("/user/{id}")
View user(
  @PathVariable("id") userId,
  @ModelAttribute("form") @Valid UserForm form,
  BindingResult bindingResult,
  RedirectAttributes redirectAttributes,
  HtmxResponse htmxResponse,
  Model model) {

  if (bindingResult.hasErrors()) {
    return HtmxView("user/form");
  }
  
  // save user
  htmxResponse.addTrigger("user-updated");
  return HtmxRedirectView("/user/list");
}
```

xhaggi replied:
Depends on what you want to do. If you want htmx to swap the response into the current target then you can use a normal redirect. If you want htmx to do a client-side redirect to a new page then you need the `HtmxRedirectView`.

wimdeblauwe replied:
I just did a test on an example from my book and this is not the behaviour I see. This is the original code:

```
   @PostMapping("/new")
    public String createNewContact(Model model,
                                   @ModelAttribute("formData") @Valid CreateContactFormData formData,
                                   BindingResult bindingResult,
                                   RedirectAttributes redirectAttributes) {
        if (bindingResult.hasErrors()) {
            model.addAttribute("editMode", EditMode.CREATE); // <.>
            return "contacts/edit";
        }

        service.storeNewContact(formData.getGivenName(),
                formData.getFamilyName(),
                formData.getPhone(),
                formData.getEmail());

        redirectAttributes.addFlashAttribute("successMessage",
                "Created New Contact!");

        return "redirect:/contacts";
    }
```

If I use the code from this branch then everything still works. The browser redirects to the `/contacts` URL and the flash message is shown.

If I adjust the code to :

```
    @PostMapping("/new")
    public Object createNewContact(Model model,
                                 @ModelAttribute("formData") @Valid CreateContactFormData formData,
                                 BindingResult bindingResult,
                                 RedirectAttributes redirectAttributes) {
        if (bindingResult.hasErrors()) {
            model.addAttribute("editMode", EditMode.CREATE); // <.>
            return new HtmxView("contacts/edit");
        }

        service.storeNewContact(formData.getGivenName(),
                formData.getFamilyName(),
                formData.getPhone(),
                formData.getEmail());

        redirectAttributes.addFlashAttribute("successMessage",
                "Created New Contact!");

        return new HtmxRedirectView("/contacts");
    }
```

(Note that I need to return `Object` since there is no common interface between `HtmxView` and `HtmxRedirectView`)

Then there is also a redirect in the browser, but the page is empty. If I check the response in DevTools, it is completely empty.

checketts replied:
Are you using a link with an href that is being followed? So you are navigating to `/new` which is cancelling the `contacts` call?

wimdeblauwe replied:
It is a button in a form like this:

```
<form
          th:object="${formData}"

          hx:post="${editMode?.name() == 'UPDATE'}?@{/contacts/{id}/edit(id=${formData.id})}:@{/contacts/new}"
          class="flex flex-col gap-y-2"> <!--.-->
        <div th:replace="~{fragments/forms :: textinput('Given Name', 'givenName')}"></div>
        <div th:replace="~{fragments/forms :: textinput('Family Name', 'familyName')}"></div>
        <div th:replace="~{fragments/forms :: textinput('Phone', 'phone')}"></div>
        <div th:replace="~{fragments/forms :: emailinput('Email', 'email', '/contacts/new')}"></div> <!--.-->
        <button type="submit" class="button-primary mt-4">Save</button>
    </form>
```

xhaggi replied:
>(Note that I need to return Object since there is no common interface between HtmxView and HtmxRedirectView)

That's completely fine.

Could you please show us a screenshot of the details (response header) for the request to `/contact` in your browser?

checketts replied:
Sounds like a `preventDefault` is needed on your button: See https://github.com/bigskysoftware/htmx/issues/2323

xhaggi replied:
@checketts I don't think you are correct here. The form element has an `hx-post` attribute, the button submits the form and htmx kicks in and handles the submission without the need to call `preventDefault`.

checketts replied:
@xhaggi Yeah I doubt it is something so simple.  I'm very curious to hear about the root cause.

wimdeblauwe replied:
The POST request has these response headers:
![CleanShot 2024-11-06 at 21 25 19@2x](https://github.com/user-attachments/assets/c00d8714-aeb3-4831-a561-f31147ce8341)
The GET `/contacts` has no response headers:

![CleanShot 2024-11-06 at 21 25 49@2x](https://github.com/user-attachments/assets/a16b9b20-1c57-4d0c-8add-b6c1ffddebe2)

xhaggi replied:
Can you access `/contacts` directly in your browser? This is what the browser wants to do.

wimdeblauwe replied:
Yes, that works fine.
![CleanShot 2024-11-06 at 21 32 49@2x](https://github.com/user-attachments/assets/f9105cc1-1bab-430b-afe1-898a1551dc24)

checketts replied:
HX-Refresh header shouldn't be there

xhaggi replied:
Why?

checketts replied:
HX-Refresh does a client side refresh, HX-Redirect changes the url. If the goal was redirecting, a refresh isn't desired.

wimdeblauwe replied:
@checketts you are right. If I change the code of the library to *not* send the `HX-Refresh` header, then it works as expected.

xhaggi replied:
Mhhh ok... I thought that this didn't matter. I added the `HX-Refresh` response header to force a reload of the page in case `location.href` doesn't. This is the case if your redirect URL contains an anchor`#` and you are currently on this URL. You can verify this behavior by doing the following in your browser console.

```js
(function () {location.href='https://github.com/wimdeblauwe/htmx-spring-boot/pull/140#'})();
````

```js
(function () {location.href='https://github.com/wimdeblauwe/htmx-spring-boot/pull/140#';location.reload()})();
````

Tested in Chrome. But ok, let me remove the `HX-Refresh` header if this fix the issue.

xhaggi replied:
Force-pushed.

xhaggi replied:
I wonder why htmx has this [code path](https://github.com/bigskysoftware/htmx/blob/master/src/htmx.js#L4653) in case both `HX-Redirect` and `HX-Refresh` are present?

xhaggi replied:
I have verified that you are correct, @checketts. The following call does not redirect, it just reloads the current page.

```js
(function () {location.href='https://github.com/wimdeblauwe/htmx-spring-boot/';location.reload()})();
```

checketts replied:
Yeah I can't think of a good reason unless the redirect was for the same page with an anchor. Like you described.

wimdeblauwe replied:
Thanks for the update. I did another test to compare using `return "redirect:/contacts"` vs `return new HtmxRedirectView("/contacts")` in a `@PostMapping`:

### Using Spring MVC redirect

In this case, when issuing the `POST` through htmx, the browser receives a 302 response status. It follows this redirect, but does this with the `HX-Request` request header set to `true`. So it goes to `/contacts` but there, the controller method (in my example) handles this as an htmx request and only returns a partial.

### Using `HtmxRedirectView`

Here, the `POST` returns a 200 response status with the `HX-Redirect: /contacts` response header. This is intercepted by htmx and the browser changes the URL and loads that page as a normal page load, not an htmx request.


I fully understand why all of this is happening, and I think it is ok. I just wonder if we should add something about this to the documentation to make it clear? I think the recommendation should be to use `return "redirect:/contacts"` if you are not using htmx in the request and use `return new HtmxRedirectView("/contacts")` for an htmx request.

xhaggi replied:
>I think the recommendation should be to use return "redirect:/contacts" if you are not using htmx in the request and use return new HtmxRedirectView("/contacts") for an htmx request.

As I said, it depends on what you want to achieve. A normal redirect is always processed by the browser, no matter if it is an AJAX request or a normal request. So if you make an htmx request and want the response to be swapped into the current target, a normal redirect is the way to do. If you want to perform a "real" page redirect, ‘htmx:redirect’ should be used. IMO this is more related to htmx and should be in the htmx documentation rather than here.

wimdeblauwe replied:
ok. Maybe something I could create a blog post about later indeed.

wimdeblauwe reviewed (commented):
Thanks again for all your hard work on this! We will need to update the README as well to document the behaviour.

It is correct that everything is still backwards compatible for now? So we can release this as 3.6.0 ?

xhaggi commented on htmx-spring-boot/src/main/java/io/github/wimdeblauwe/htmx/spring/boot/mvc/HtmxHandlerMethodArgumentResolver.java:
```diff
@@ -12,47 +14,22 @@
     @Override
     public boolean supportsParameter(MethodParameter parameter) {
-        return parameter.getParameterType().equals(HtmxRequest.class);
+        return parameter.getParameterType() == HtmxRequest.class;
```
The preferred way to compare classes is `==`. Spring does the same in their resolvers.

xhaggi commented:
>Thanks again for all your hard work on this!

I'd be glad to. 😊

>We will need to update the README as well to document the behaviour.

Will adjust the README tomorrow.

>It is correct that everything is still backwards compatible for now?

Yep, that's right.

xhaggi commented:
I have made another force-push to tackle some issues that I have found myself. The `HtmxLocationRedirectView` was missing some constructor overloads, I added `@since 3.6.0` to most of the new stuff and finally I added a new commit that introduces a `HtmxViewResolver` to handle special view name resolutions for the redirect and refresh views.

Now you can write something this:
```java
@PostMapping("/user/{id}")
String saveUser() {
  if (error) {
    return "/user/form";
  }
  // success
  return "htmx:redirect:/user/list";
}
```
The following special view name prefixes are supported:
* `htmx:redirect:/path` -> return a `HtmxRedirectView` with the given path
* `htmx:location:/path` -> return a `HtmxLocationRedirectView` with the given path
* `htmx:refresh` -> return a `HtmxRefreshView`
